src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1251 @@
+							   "database")));
+		exit(0);
+	}
+
+	/* make worker recognizable in pg_stat_activity */
+	pgstat_report_appname("citus background task queue monitor");
+
+	ereport(DEBUG1, (errmsg("started citus background task queue monitor")));
+
+	/* TODO this is here for debugging purposses, remove before merge. */
+	if (BackgroundTaskMonitorDebugDelay)
+	{
+		pg_usleep(30 * 1000 * 1000);
+	}
+
+	MemoryContext perTaskContext = AllocSetContextCreate(CurrentMemoryContext,
+														 "PerTaskContext",
+														 ALLOCSET_DEFAULT_MINSIZE,
+														 ALLOCSET_DEFAULT_INITSIZE,
+														 ALLOCSET_DEFAULT_MAXSIZE);
+
+	/*
+	 * First we find all jobs that are running, we need to check if they are still running
+	 * if not reset their state back to scheduled.
+	 */
+	{
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		ResetRunningBackgroundTasks();
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+	}
+
+
+	MemoryContext oldContextPerJob = MemoryContextSwitchTo(perTaskContext);
+	bool hasJobs = true;
+	while (hasJobs)
+	{
+		MemoryContextReset(perTaskContext);
+
+		CHECK_FOR_INTERRUPTS();
+
+		InvalidateMetadataSystemCache();
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		/*
+		 * We need to load the task into the perTaskContext as we will switch contexts
+		 * later due to the committing and starting of new transactions
+		 */
+		MemoryContext oldContext = MemoryContextSwitchTo(perTaskContext);
+		BackgroundTask *task = GetRunnableBackgroundTask();
+
+		if (!task)
+		{
+			MemoryContextSwitchTo(oldContext);
+
+			PopActiveSnapshot();
+			CommitTransactionCommand();
+
+			hasJobs = false;
+			break;
+		}
+
+		/* we load the database name and username here as we are still in a transaction */
+		char *databaseName = get_database_name(MyDatabaseId);
+		char *userName = GetUserNameFromId(task->owner, false);
+
+		MemoryContextSwitchTo(oldContext);
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+
+		MemoryContextSwitchTo(perTaskContext);
+
+		/*
+		 * The background worker needs to be started outside of the transaction, otherwise
+		 * it will complain about leaking shared memory segments used, among other things,
+		 * to communicate the output of the backend.
+		 */
+		dsm_segment *seg = NULL;
+		BackgroundWorkerHandle *handle =
+			StartCitusBackgroundJobExecuter(databaseName, userName, task->command,
+											task->taskid, &seg);
+
+		if (handle == NULL)
+		{
+			ereport(ERROR, (errmsg("unable to start background worker for background "
+								   "task execution")));
+		}
+
+		pid_t pid = 0;
+		GetBackgroundWorkerPid(handle, &pid);
+
+		ereport(LOG, (errmsg("found task with jobid/taskid: %ld/%ld",
+							 task->jobid, task->taskid)));
+
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		task->status = BACKGROUND_TASK_STATUS_RUNNING;
+		SET_NULLABLE_FIELD(task, pid, pid);
```

Comment:
there is a scenario where the monitor fails before updating the pid (e.g. it errors with "tuple concurrently updated")

in that case, we would not have registered the pid anywhere, and we would not be able to kill it (ResetRunningBackgroundTasks skips it, since pid is null)

Reply:
potentially, when users update the tuple themself after they have been committed when the job was created. Although I think the timing is very strict, between finding the task and spawning the backend.

It will be easiest with a debugger attached. I suspect it finding lack of locking, or the fact that it loses the first HeapTuple before updating, meaning that it would update the tuple with the full state after the user would have updated it.

Reply:
the case of concurrent updates with `citus_job_cancel` have been addressed by using an `AccessExclusiveLock` on `pg_dist_background_task` while updating any task.

We also re-read the tasks at those moment to not lose concurrent updates that have happened outside of the monitor. This covers most cases described here. Still not the one where we kill the executors in the shutdown hook. That would be a great improvement later.